### PR TITLE
reside-77: Start on translation lintr

### PR DESCRIPTION
**Pull request by richfitz**
Apologies for the size of this one, though not sure where it can be usefully split.

This PR implements the first pass at a translation lintr. Lots of bits to add here, but I think the core is OK.

Give it a go on hintr with:

```r
devtools::load_all()
path <- "~/Documents/Projects/hint/hintr" # adjust as you need
devtools::load_all(path)
traduire::lint_translations_report(lint_translations_package(path))
```

### CODE REVIEW

File: R/lint.R [hunk 1]
```diff
@@ -0,0 +1,279 @@
...
+##'
+##' @param file File to save the report in (default is to use a
+##'   temporary file).
+##'
+##' @return This function is called for the side effect of viewing a
+##'   report.  It will return (invisibly) the path to the produced
+##'   report.
+##'
+##' @export
+lint_translations_report <- function(x, file = tempfile()) {
```
Comment (r-ash) [inline R/lint.R]: We discussed this in person but we should put `.html` extension on tempfile name so chrome renders the page nicely. At the moment it shows the raw html.

File: R/lint.R [hunk 2]
```diff
@@ -0,0 +1,279 @@
...
+
+  ## This is the overall expression id - if this is highlighted then
+  ## the whole relevant part of the translation will be.
+  ##
+  ## The more robust way of getting this looks to be to get the parent
+  ## of this, but we're working with basically undocumented data
+  ## structures here...
+  index <- data$index[[i - 1L]]
+
+  res <- parse_data_match_call(i, data, function(key, data, ...) {})
```
Comment (r-ash) [inline R/lint.R]: I had some confusion with this. For me `res` and function name `parse_data_match_call` don't have much meaning. I wonder if a naming more in the domain of the problem we're trying to solve would be easier to read? Something like
`args <- get_function_args(index, data, order = c("key", "data", "..."))` would be clearer. Though definitely possible that just stems from unfamiliarity with `getParseData` and `match.call`. Or maybe some roxygen docs on `parse_data_match_call` itself about what the function trying to achieve would be useful.

File: R/string.R [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+    loc <- str_find(paste0(prefix, s, suffix), text, pos)
+    pos <<- loc[[2]]
+    found <<- c(found,
+                list(list(text = trimws(s), from = loc[[1]], to = loc[[2]])))
+    s
+  }
+  glue::glue(text,
+             .transformer = extractor,
+             .open = prefix, .close = suffix)
+  data_frame(text = vcapply(found, "[[", "text"),
```
Comment (r-ash) [inline R/string.R:14]: Nice trick with the `vapply` `"[["`. That will come in handy!

File: R/string.R [hunk 4]
```diff
@@ -0,0 +1,158 @@
...
+  i <- !is.na(msg)
+  messages <- data_frame(tag = tag[i], line = line[i], value = msg[i])
+
+  if (filter) {
+    lines <- min(line):max(line)
+    text <- text[lines]
+    messages$index <- messages$line - lines[[1]] + 1L
+  } else {
+    lines <- seq_along(text)
+    ## messages$index <- lines
```
Comment (r-ash) [inline R/string.R]: Should this have been removed or still needed?

File: tests/testthat/test-string.R [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+})
+
+
+test_that("str_insert", {
+  expect_equal(str_insert("my string", 3, " amazing"), "my amazing string")
+})
+
+
+test_that("html escape", {
+  ##         123356
```
Comment (r-ash) [inline tests/testthat/test-string.R]: Another spurious comment?

File: R/string.R [hunk 6]
```diff
@@ -0,0 +1,158 @@
...
+    render = function(tags, filter = TRUE, group = NULL, escape = FALSE) {
+      markup_render(tags, self$spans, self$text, filter, group, escape)
+    },
+
+    summary = function() {
+      tags <- lint_tags_names()
+      spans <- matrix(self$spans, 2)
+      m <- matrix(FALSE, length(self$text), length(tags),
+                  dimnames = list(NULL, tags))
+      ## count <- as.list(table(factor(vcapply(spans[1, ], "[[", "tag"), tags)))
```
Comment (r-ash) [inline R/string.R]: A line for removal?

File: inst/report/script.js [hunk 7]
```diff
@@ -0,0 +1,23 @@
...
+    for (var i = 0; i < tabcontent.length; i++) {
+        tabcontent[i].style.display = "none";
+    }
+
+    var tablinks = document.getElementsByClassName("tablinks");
+    for (var i = 0; i < tablinks.length; i++) {
+        tablinks[i].className = tablinks[i].className.replace(" active", "");
+    }
+
+    // tabcontent.forEach((el) => {
```
Comment (r-ash) [inline inst/report/script.js]: These lines commented out too?

**Review by r-ash — APPROVED** [review]
Nice stuff. Lots of new things for me here so not much to say given how long I have spent looking at this. I think only one real thing to change which is to put `html` file extension on tempfile so it renders in chrome. We should make tickets for the TODOs too and add the task number to the notes.

File: tests/testthat/test-string.R (hunk 5 above)
Comment (richfitz) [reply tests/testthat/test-string.R]: This time no, it's column number. Added a comment for the comment 5582e60

File: R/string.R (hunk 4 above)
Comment (richfitz) [reply R/string.R]: Removed in 2791f87

File: R/lint.R (hunk 2 above)
Comment (richfitz) [reply R/lint.R]: Expanded a bit in dababc7 and renamed `res` to `args` to be a bit clearer - does that help?

File: R/string.R (hunk 6 above)
Comment (richfitz) [reply R/string.R]: Removed in 2791f87

File: inst/report/script.js (hunk 7 above)
Comment (richfitz) [reply inst/report/script.js]: Thanks - was an alternative implementation attempt though it did no work and I'm not sure it's any better. Removed in 5eeda01

File: tests/testthat/test-string.R (hunk 5 above)
Comment (richfitz) [reply tests/testthat/test-string.R]: This time no, it's column number. Added a comment for the comment 5582e60

Comment (r-ash) [reply tests/testthat/test-string.R]: `a < 1 && b > 2`

Nice example.

**Review by r-ash — APPROVED** [review]

**Pull request merged**